Templates/Templates/ModelMappable.stencil

```diff
@@ -0,0 +1,162 @@
+private struct Formatter {
+  static let iso8601: DateFormatter = {
+    let formatter = DateFormatter()
+    formatter.calendar = Calendar(identifier: .iso8601)
+    formatter.locale = Locale(identifier: "en_US_POSIX")
+    formatter.timeZone = TimeZone(abbreviation: "UTC")
+    formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
+    return formatter
+  }()
+}
+
+extension String {
+  fileprivate var dateFromISO8601: Date? {
+    // "Mar 22, 2017, 10:22 AM"
+    return Formatter.iso8601.date(from: self)
+  }
+}
+
+extension Date {
+  fileprivate var iso8601: String {
+    return Formatter.iso8601.string(from: self)
+  }
+
+  fileprivate static func fromTimestamp(_ timestamp: Timestamp) -> Date {
+    return Date(timeIntervalSince1970: Double(timestamp) / 1000)
+  }
+
+  fileprivate var asTimestamp: Timestamp {
+    return UInt64(1000 * timeIntervalSince1970)
+  }
+}
+
+{% macro modelName name %}{{ name | replace:"Mapping","Model" }}{% endmacro %}
+{% macro remoteName var %}{% if var.annotations.coding %}{{ var.annotations.coding }}{% elif var.typeName.name == "String" %}{{ var.name|camelToSnakeCase|replace:"id","uuid" }}{% else %}{{ var.name|camelToSnakeCase }}{% endif %}{% endmacro %}
+
+{% macro immutableVariables variables %}
+  {% for var in variables %}
+  let {{ var.name }}: {{ var.typeName }}
+  {% endfor %}
+{% endmacro %}
+
+{% macro codingKeys variables %}
+  private enum CodingKeys: String, CodingKey {
+    {% for var in variables %}
+    case {{ var.name }} = "{% call remoteName var %}"
+    {% endfor %}
+  }
+{% endmacro %}
+
+{% macro decodeVar var %}
+  {% set itype %}{{ var.annotations.incomingType }}{% endset %}
+  {% if var.typeName.name == "Bool" and itype == "Int" %} {# its a 0 or 1 #}
+    {{ var.name }} = try container.decodeIfPresent(Int.self, forKey: .{{ var.name }}) == 1
+  {% elif var.typeName.name == "Date" and itype == "String" %} {# iso date string #}
+    if
+      let {{ var.name }} = try container.decodeIfPresent(String.self, forKey: .{{ var.name }}),
+      let date = {{ var.name }}.dateFromISO8601 {
+      self.{{ var.name }} = date
+    } else {
+      self.{{ var.name }} = Date()
+    }
+  {% endif %}
+{% endmacro %}
+
+{# cant get spacing right without doing it like this, better way when possible #}
+{% macro decodeWithDefault var %}
+  {% set typename %}{{ var.typeName.name }}{% endset %}
+  {% if typename == "String" %}
+    {{ var.name }} = try container.decodeIfPresent({{ typename }}.self, forKey: .{{ var.name }}) ?? ""
+  {% elif typename == "Bool" %}
+    {{ var.name }} = try container.decodeIfPresent({{ typename }}.self, forKey: .{{ var.name }}) ?? false
+  {% elif typename == "Int" or typename == "Double" or typename == "Float" or typename == "NSNumber" %}
+    {{ var.name }} = try container.decodeIfPresent({{ typename }}.self, forKey: .{{ var.name }}) ?? 0
+  {% elif typename == "Date" %}
+    {{ var.name }} = try container.decodeIfPresent({{ typename }}.self, forKey: .{{ var.name }}) ?? Date()
```

Comment:
not sure that current date is a good default. we could use annotation to specify default value and extend this macros to support any variable annotated with it.